controller_interface/include/semantic_components/imu_sensor.hpp

```diff
@@ -27,110 +27,99 @@ namespace semantic_components
+    std::array<double, 4> orientation;
+    for (auto i = 0u; i < orientation.size(); ++i)
     {
-      orientation_[i] = state_interfaces_[interface_offset + i].get().get_value();
+      orientation[i] = state_interfaces_[i].get().get_value();
     }
-    return orientation_;
+    return orientation;
   }
 
   /// Return angular velocity.
   /**
    * Return angular velocity reported by an IMU
    *
-   * \return array of size 3 with angular velocity values.
+   * \return array of size 3 with angular velocity values (x, y, z).
    */
-  std::array<double, 3> get_angular_velocity()
+  std::array<double, 3> get_angular_velocity() const
   {
-    size_t interface_offset = orientation_.size();
-    for (size_t i = 0; i < angular_velocity_.size(); ++i)
+    std::array<double, 3> angular_velocity;
+    const std::size_t interface_offset{4};
+    for (auto i = 0u; i < angular_velocity.size(); ++i)
     {
-      angular_velocity_[i] = state_interfaces_[interface_offset + i].get().get_value();
+      angular_velocity[i] = state_interfaces_[interface_offset + i].get().get_value();
     }
-    return angular_velocity_;
+    return angular_velocity;
   }
 
   /// Return linear acceleration.
   /**
    * Return linear acceleration reported by an IMU
    *
-   * \return array of size 3 with linear acceleration values.
+   * \return array of size 3 with linear acceleration values (x, y, z).
    */
-  std::array<double, 3> get_linear_acceleration()
+  std::array<double, 3> get_linear_acceleration() const
   {
-    size_t interface_offset = orientation_.size() + angular_velocity_.size();
-    for (size_t i = 0; i < linear_acceleration_.size(); ++i)
+    std::array<double, 3> linear_acceleration;
+    const std::size_t interface_offset{7};
+    for (auto i = 0u; i < linear_acceleration.size(); ++i)
     {
-      linear_acceleration_[i] = state_interfaces_[interface_offset + i].get().get_value();
+      linear_acceleration[i] = state_interfaces_[interface_offset + i].get().get_value();
     }
-    return linear_acceleration_;
+    return linear_acceleration;
   }
 
   /// Return Imu message with orientation, angular velocity and linear acceleration
   /**
    * Constructs and return a IMU message from the current values.
    * \return imu message from values;
    */
-  bool get_values_as_message(sensor_msgs::msg::Imu & message)
+  bool get_values_as_message(sensor_msgs::msg::Imu & message) const
   {
-    // call get_orientation() and get_angular_velocity()  get_linear_acceleration() to
-    // update with the latest values
-    get_orientation();
-    get_angular_velocity();
-    get_linear_acceleration();
+    const auto [orientation_x, orientation_y, orientation_z, orientation_w] = get_orientation();
+    const auto [angular_velocity_x, angular_velocity_y, angular_velocity_z] =
+      get_angular_velocity();
+    const auto [linear_acceleration_x, linear_acceleration_y, linear_acceleration_z] =
+      get_linear_acceleration();
```

Comment:
```suggestion
    const auto [message.orientation.x, message.orientation.y, message.orientation.z, message.orientation.w] = get_orientation();
    const auto [message.angular_velocity.x, message.angular_velocity.y, message.angular_velocity.z] =
      get_angular_velocity();
    const auto [message.linear_acceleration.x, message.linear_acceleration.y, message.linear_acceleration.z] =
      get_linear_acceleration();
```

Same here

Reply:
Unfortunately it will not work, but compiler should optimize away that local variables.

Reply:
Well we have return value optimization in those cases. SO, the compiler does the job for you. I agree that it needs to be const ref return

Reply:
You can use `std::tie` for this https://en.cppreference.com/w/cpp/utility/tuple/tie

Reply:
Unluckily, it would work with struct or tuple. It won't work in array case.

Reply:
You are right.